File: 3rd-semester/object-oriented-programming/containers/java/src/main/java/com/university/oop/containers/service/Ship.java
package com.university.oop.containers.service;

import com.university.oop.containers.model.Container;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.logging.Logger;

/**
 * Represents a cargo ship with a fixed maximum container capacity.
 *
 * <p>Manages a fleet of {@link Container} objects, computes the total
 * shipping cost, and supports listing all loaded containers.</p>
 */
public class Ship {

    private static final Logger LOGGER = Logger.getLogger(Ship.class.getName());

    private final int capacity;
    private final List<Container> containers = new ArrayList<>();

    /**
     * Constructs a Ship.
     *
     * @param capacity maximum number of containers this ship can carry
     */
    public Ship(int capacity) {
        this.capacity = capacity;
    }

    /** @return maximum container capacity */
    public int getCapacity() { return capacity; }

    /** @return unmodifiable view of all loaded containers */
    public List<Container> getContainers() {
        return Collections.unmodifiableList(containers);
    }

    /**
     * Loads a container onto the ship if capacity allows.
     *
     * @param container the container to load
     */
    public void addContainer(Container container) {
        if (containers.size() < capacity) {
            containers.add(container);
        } else {
            LOGGER.warning("Ship at full capacity (" + capacity + "). Cannot add: " + container.getCode());
        }
    }

    /**
     * Calculates the total shipping cost for all loaded containers.
     *
     * @return total cost
     */
    public double getTotalCost() {
        return containers.stream()
                .mapToDouble(Container::getCost)
                .sum();
    }

    /** Logs the details of all loaded containers. */
    public void printContainers() {
        containers.forEach(Container::printInfo);
    }
}

File: 3rd-semester/object-oriented-programming/containers/csharp/src/Services/Ship.cs
using Microsoft.Extensions.Logging;
using ShipCargoSystem.Models;

namespace ShipCargoSystem.Services;

/// <summary>
/// Represents a cargo ship with a fixed maximum container capacity.
/// </summary>
public class Ship
{
    private readonly ILogger<Ship>   _logger;
    private readonly List<Container> _containers = [];
    private readonly int             _capacity;

    /// <summary>Gets all loaded containers (read-only).</summary>
    public IReadOnlyList<Container> Containers => _containers.AsReadOnly();

    /// <summary>Initialises the Ship.</summary>
    /// <param name="capacity">Maximum number of containers.</param>
    /// <param name="logger">Logger instance.</param>
    public Ship(int capacity, ILogger<Ship> logger)
    {
        _capacity = capacity;
        _logger   = logger;
    }

    /// <summary>
    /// Loads a container if capacity allows; warns and skips otherwise.
    /// </summary>
    /// <param name="container">The container to load.</param>
    public void AddContainer(Container container)
    {
        if (_containers.Count < _capacity)
            _containers.Add(container);
        else
            _logger.LogWarning("Ship at full capacity ({Cap}). Cannot add {Code}.", _capacity, container.Code);
    }

    /// <summary>Calculates the total shipping cost for all loaded containers.</summary>
    /// <returns>Total cost.</returns>
    public double GetTotalCost() => _containers.Sum(c => c.GetCost());

    /// <summary>Logs all loaded containers.</summary>
    public void PrintContainers()
    {
        foreach (var c in _containers)
            _logger.LogInformation("{Container}", c);
    }
}
